File: hgnfs_py/core/lorentz.py
"""
Lorentz (hyperboloid) manifold operations.
Model: L^d = { x ∈ R^{d+1} : ⟨x, x⟩_L = -1, x_0 > 0 }

All operations follow the HGCN reference implementation (refs/hgcn) with 
the Euclidean parametrization trick from the Numerical Stability paper.
"""

import torch
import torch.nn.functional as F
from hgnfs_py.core.numerical import COSH_SINH_CLIP, NEAR_ZERO_EPS, ARCCOSH_TOL, safe_arcosh, clamp_norm

#  origin (d+1 dimensional) 
_ORIGIN_DIMS = {}  # cache per dimension


def origin(dim: int) -> torch.Tensor:
    """Lorentz origin o = (1, 0, ..., 0) ∈ L^d. d = spatial dim."""
    if dim not in _ORIGIN_DIMS:
        o = torch.zeros(dim + 1)
        o[0] = 1.0
        _ORIGIN_DIMS[dim] = o
    return _ORIGIN_DIMS[dim].clone()


LORENTZ_ORIGIN = (1.0,)  # scalar sentinel


#  inner product / norm 

def lorentz_inner(x: torch.Tensor, y: torch.Tensor, keepdim: bool = False) -> torch.Tensor:
    """
    ⟨x, y⟩_L = -x_0·y_0 + Σ_{i=1}^{d} x_i·y_i

    Args:
        x, y: (*, d+1)  - batch of Lorentz points
    Returns:
        (*, 1) if keepdim else (*,)
    """
    xy = x * y
    inner = xy[..., 1:].sum(dim=-1) - xy[..., 0]
    if keepdim:
        inner = inner.unsqueeze(-1)
    return inner


def lorentz_norm(x: torch.Tensor, keepdim: bool = False) -> torch.Tensor:
    """‖v‖_L = √⟨v, v⟩_L (used for tangent vectors)."""
    dot = lorentz_inner(x, x, keepdim=keepdim)
    return torch.sqrt(torch.clamp(dot, min=NEAR_ZERO_EPS))


#  geodesic distance 

def distance(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
    """
    d_L(x, y) = arcosh(-⟨x, y⟩_L)

    Args:
        x, y: (*, d+1)  - Lorentz points on L^d
    Returns:
        (*,)  - pairwise geodesic distances
    """
    inner = lorentz_inner(x, y, keepdim=False)
    arg = torch.clamp(-inner, min=1.0 + ARCCOSH_TOL)
    return safe_arcosh(arg)


def squared_distance(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
    """d²(x, y) = arcosh(-⟨x, y⟩_L)²  (squared geodesic distance). Clamped to 50 per HGCN."""
    d = distance(x, y)
    sq = d * d
    return torch.clamp(sq, max=50.0)


#  exponential map 

def exp_o(v: torch.Tensor) -> torch.Tensor:
    """
    Exponential map from the origin.
    v: (*, d+1) - tangent vector at origin (v_0 is ignored, must be 0).
    Returns: (*, d+1) - point on Lorentz manifold.
    """
    v_space = v[..., 1:]  # (*, d)
    v_norm = torch.norm(v_space, p=2, dim=-1, keepdim=True)  # (*, 1)
    v_norm = torch.clamp(v_norm, min=NEAR_ZERO_EPS)
    theta = clamp_norm(v_norm)

    res = torch.zeros_like(v)
    res[..., 0] = torch.cosh(theta).squeeze(-1)
    direction = v_space / v_norm
    res[..., 1:] = torch.sinh(theta) * direction
    return _project(res)


def exp_x(v: torch.Tensor, x: torch.Tensor) -> torch.Tensor:
    """
    Exponential map from point x.
    v: (*, d+1) - tangent vector at x.
    x: (*, d+1) - base point on Lorentz manifold.
    Returns: (*, d+1) - exp_x(v).
    """
    v_norm = lorentz_norm(v, keepdim=True)
    v_norm = torch.clamp(v_norm, max=1e6)
    theta = torch.clamp(v_norm, min=NEAR_ZERO_EPS)
    cosh_t = torch.cosh(theta)
    sinh_t = torch.sinh(theta)
    result = cosh_t * x + sinh_t * v / theta
    return _project(result)


#  logarithmic map 

def log_o(x: torch.Tensor) -> torch.Tensor:
    """
    Logarithmic map to the origin.
    x: (*, d+1) - point on Lorentz manifold.
    Returns: (*, d+1) - tangent vector at origin (v_0 = 0).
    """
    x_space = x[..., 1:]  # (*, d)
    x_norm = torch.norm(x_space, p=2, dim=-1, keepdim=True)
    x_norm = torch.clamp(x_norm, min=NEAR_ZERO_EPS)
    theta = torch.clamp(x[..., 0:1], min=1.0 + ARCCOSH_TOL)
    dist = safe_arcosh(theta)
    res = torch.zeros_like(x)
    res[..., 1:] = dist * (x_space / x_norm)
    # time component stays 0 (tangent at origin)
    return res


def log_x(y: torch.Tensor, x: torch.Tensor) -> torch.Tensor:
    """
    Logarithmic map from x to y.
    x, y: (*, d+1) - points on Lorentz manifold.
    Returns: (*, d+1) - tangent vector at x pointing toward y.
    """
    # ⟨x, y⟩_L + 1 ≤ 0 for points on hyperboloid (Cauchy-Schwarz)
    xy = lorentz_inner(x, y, keepdim=True) + 1.0
    xy = torch.clamp(xy, max=-ARCCOSH_TOL) - 1.0
    u = y + xy * x
    u_norm = lorentz_norm(u, keepdim=True)
    u_norm = torch.clamp(u_norm, min=NEAR_ZERO_EPS)
    dist = distance(x, y).unsqueeze(-1)
    result = dist * u / u_norm
    return _project_tan(result, x)


#  parallel transport 

def parallel_transport_o_to_x(v: torch.Tensor, x: torch.Tensor) -> torch.Tensor:
    """
    Parallel transport of tangent vector v from origin to x.
    v: (*, d+1) - tangent vector at origin (v_0 = 0).
    x: (*, d+1) - destination point.
    Returns: (*, d+1) - tangent vector at x.
    """
    x0, xs = x[..., 0:1], x[..., 1:]
    xs_norm = torch.norm(xs, p=2, dim=-1, keepdim=True)
    xs_norm = torch.clamp(xs_norm, min=NEAR_ZERO_EPS)
    xs_dir = xs / xs_norm

    # Direction vector from HGCN ptransp0 formula
    dir_0 = -xs_norm
    dir_s = (1.0 - x0) * xs_dir
    alpha = (v[..., 1:] * xs_dir).sum(dim=-1, keepdim=True)
    result = v - alpha * torch.cat([dir_0, dir_s], dim=-1)
    return _project_tan(result, x)


#  manifold utilities 

def _project(x: torch.Tensor) -> torch.Tensor:
    """
    Project any vector onto the Lorentz manifold by replacing x₀ with
    sqrt(1 + ‖x_space‖²). Note: since cosh²θ - sinh²θ ≠ 1 in float32,
    this creates a small perturbation (~1e-6) from the exact exp_o result.
    Used only for truly unconstrained inputs, not after exp_o/exp_x.
    """
    xs = x[..., 1:]
    xs_sqnorm = (xs * xs).sum(dim=-1, keepdim=True)
    x0 = torch.sqrt(torch.clamp(1.0 + xs_sqnorm, min=NEAR_ZERO_EPS))
    return torch.cat([x0, xs], dim=-1)


def _project_tan(v: torch.Tensor, x: torch.Tensor) -> torch.Tensor:
    """Project v onto the tangent space at x: T_x L^d."""
    v_space = v[..., 1:]
    x_space = x[..., 1:]
    # ⟨x, v⟩_L should be 0 for tangent vectors
    vx = (x_space * v_space).sum(dim=-1, keepdim=True)
    v0 = vx / torch.clamp(x[..., 0:1], min=NEAR_ZERO_EPS)
    return torch.cat([v0, v_space], dim=-1)


def is_on_manifold(x: torch.Tensor, eps: float = 1e-4) -> torch.Tensor:
    """
    Check if points satisfy Lorentz constraint: ⟨x,x⟩_L = -1 and x₀ > 0.
    Uses relative tolerance scaled by ‖x‖² to handle large-magnitude points
    where float32 can't verify the constraint to absolute 1e-4.
    """
    inner = lorentz_inner(x, x, keepdim=False)
    # Relative tolerance: allow error proportional to embedding magnitude
    scale = torch.clamp(x[..., 0] ** 2, min=1.0)
    constraint_ok = torch.abs(inner + 1.0) < (eps * scale)
    time_ok = x[..., 0] > 0
    return constraint_ok & time_ok


#  batch conversion 

def euclidean_to_lorentz(z: torch.Tensor) -> torch.Tensor:
    """
    Euclidean parametrization z ∈ R^d -> Lorentz point x ∈ L^d.
    Uses exp_o under the hood. z_norm clamped to COSH_SINH_CLIP.
    
    Args: z: (*, d) - Euclidean parameters
    Returns: (*, d+1) - Lorentz points
    """
    v = F.pad(z, (1, 0), value=0.0)  # prepend time=0
    return exp_o(v)


def lorentz_to_euclidean(x: torch.Tensor) -> torch.Tensor:
    """Lorentz point -> Euclidean parametrization (log_o, discard time)."""
    return log_o(x)[..., 1:]


#  Möbius matrix-vector multiplication 

def mobius_matvec(W: torch.Tensor, x: torch.Tensor) -> torch.Tensor:
    """
    Hyperbolic linear transform: W ⊗ x = exp_o(W · log_o(x)).
    W: (out_dim, in_dim) - weight matrix.
    x: (*, in_dim+1) - Lorentz points.
    Returns: (*, out_dim+1) - transformed Lorentz points.
    """
    u = log_o(x)  # (*, in_dim+1)
    u_space = u[..., 1:]  # (*, in_dim)
    mu = F.linear(u_space, W)  # (*, out_dim)
    mu_padded = F.pad(mu, (1, 0), value=0.0)
    return _project(exp_o(mu_padded))


def einstein_midpoint(
    points: torch.Tensor, weights: torch.Tensor | None = None
) -> torch.Tensor:
    """
    Weighted Einstein midpoint (hyperbolic barycenter).
    points: (N, d+1) - Lorentz points.
    weights: (N,) or None (uniform) - positive weights.
    Returns: (d+1,) - the weighted centroid.

    Algorithm: Weighted sum in Klein coordinates, weighted by Lorentz gamma.
    Per HypRAG's Outward Einstein Midpoint.
    """
    x0, xs = points[..., 0], points[..., 1:]  # (N,), (N, d)
    # Klein coordinates: k = x_space / x_time
    k = xs / torch.clamp(x0.unsqueeze(-1), min=NEAR_ZERO_EPS)  # (N, d)
    k_norm2 = (k * k).sum(dim=-1)  # (N,)
    # Lorentz gamma factor
    gamma = 1.0 / torch.sqrt(torch.clamp(1.0 - k_norm2, min=NEAR_ZERO_EPS))  # (N,)
    if weights is None:
        weights = torch.ones_like(gamma)
    phi = weights * gamma  # outward re-weighting
    w_tilde = phi / torch.clamp(phi.sum(), min=NEAR_ZERO_EPS)
    num = (w_tilde.unsqueeze(-1) * gamma.unsqueeze(-1) * k).sum(dim=0)
    denom = (w_tilde * gamma).sum()
    m_k = num / torch.clamp(denom, min=NEAR_ZERO_EPS)
    m_k_norm2 = (m_k * m_k).sum()
    denom2 = torch.sqrt(torch.clamp(1.0 - m_k_norm2, min=NEAR_ZERO_EPS))
    x0_out = 1.0 / denom2
    xs_out = m_k / denom2
    return torch.cat([x0_out.unsqueeze(0), xs_out])

File: src/zetla/rag/core/lorentz.cpp
#include "lorentz.hpp"
#include <algorithm>
#include <cmath>
#include <cstring>
#include <vector>

namespace hgnfs::lorentz {

//  inner product 

float inner(const LorentzPoint& a, const LorentzPoint& b) {
    float sum = -a.t * b.t;
    int d = static_cast<int>(a.x.size());
    for (int i = 0; i < d; ++i) sum += a.x[i] * b.x[i];
    return sum;
}

float norm(const LorentzPoint& v) {
    float dot = inner(v, v);
    return (dot < NEAR_ZERO_EPS) ? 0.0f : std::sqrt(dot);
}

//  distance 

float distance(const LorentzPoint& a, const LorentzPoint& b) {
    float in = inner(a, b);
    float arg = -in;
    if (arg < 1.0f + ARCCOSH_TOL) arg = 1.0f + ARCCOSH_TOL;
    return numerical::safe_arcosh(arg);
}

float squared_distance(const LorentzPoint& a, const LorentzPoint& b) {
    float d = distance(a, b);
    float sq = d * d;
    return (sq > SQUARED_DIST_CLAMP) ? SQUARED_DIST_CLAMP : sq;
}

//  exponential map 

LorentzPoint exp_o(const float* z, int dim) {
    float z_norm = 0.0f;
    for (int i = 0; i < dim; ++i) z_norm += z[i] * z[i];
    z_norm = std::sqrt(z_norm);
    if (z_norm < NEAR_ZERO_EPS) {
        LorentzPoint o(dim);
        o.t = 1.0f;
        return o;
    }
    float theta = numerical::clamp_norm(z_norm);
    float ch = std::cosh(theta);
    float sh = std::sinh(theta);
    LorentzPoint p(dim);
    p.t = ch;
    for (int i = 0; i < dim; ++i) p.x[i] = sh * z[i] / z_norm;
    return project(p);
}

LorentzPoint exp_x(const LorentzPoint& x, const LorentzPoint& v_tan) {
    int d = static_cast<int>(x.x.size());
    float vn = norm(v_tan);
    if (vn < NEAR_ZERO_EPS) return x;
    float ch = std::cosh(vn);
    float sh = std::sinh(vn);
    LorentzPoint r(d);
    r.t = ch * x.t + sh * v_tan.t / vn;
    for (int i = 0; i < d; ++i)
        r.x[i] = ch * x.x[i] + sh * v_tan.x[i] / vn;
    return project(r);
}

//  logarithmic map 

void log_o(const LorentzPoint& x, float* z_out, int dim) {
    float xs_norm = 0.0f;
    for (int i = 0; i < dim; ++i) xs_norm += x.x[i] * x.x[i];
    xs_norm = std::sqrt(xs_norm);
    if (xs_norm < NEAR_ZERO_EPS) {
        std::memset(z_out, 0, dim * sizeof(float));
        return;
    }
    float arg = x.t;
    if (arg < 1.0f + ARCCOSH_TOL) arg = 1.0f + ARCCOSH_TOL;
    float dist = numerical::safe_arcosh(arg);
    for (int i = 0; i < dim; ++i) z_out[i] = dist * x.x[i] / xs_norm;
}

LorentzPoint log_x(const LorentzPoint& x, const LorentzPoint& y) {
    int d = static_cast<int>(x.x.size());
    float in = inner(x, y) + 1.0f;
    if (in > -ARCCOSH_TOL) in = -ARCCOSH_TOL;
    in -= 1.0f;

    LorentzPoint u(d);
    u.t = y.t + in * x.t;
    for (int i = 0; i < d; ++i) u.x[i] = y.x[i] + in * x.x[i];

    float un = norm(u);
    if (un < NEAR_ZERO_EPS) {
        LorentzPoint zero(d);
        zero.t = 0.0f;
        return zero;
    }
    float dist = distance(x, y);
    LorentzPoint v(d);
    v.t = dist * u.t / un;
    for (int i = 0; i < d; ++i) v.x[i] = dist * u.x[i] / un;
    return v;
}

//  parallel transport 

LorentzPoint parallel_transport_o_to_x(const LorentzPoint& v, const LorentzPoint& x) {
    int d = static_cast<int>(x.x.size());
    float xs_norm = 0.0f;
    for (int i = 0; i < d; ++i) xs_norm += x.x[i] * x.x[i];
    xs_norm = std::sqrt(xs_norm);
    if (xs_norm < NEAR_ZERO_EPS) return v;

    float alpha = 0.0f;
    for (int i = 0; i < d; ++i) alpha += v.x[i] * x.x[i] / xs_norm;

    LorentzPoint r(d);
    r.t = v.t - alpha * (-xs_norm);
    for (int i = 0; i < d; ++i)
        r.x[i] = v.x[i] - alpha * ((1.0f - x.t) * x.x[i] / xs_norm);
    return r;
}

//  project onto manifold 

LorentzPoint project(const LorentzPoint& x) {
    int d = static_cast<int>(x.x.size());
    float xs_sq = 0.0f;
    for (int i = 0; i < d; ++i) xs_sq += x.x[i] * x.x[i];
    float x0 = std::sqrt(1.0f + xs_sq);
    if (x0 < NEAR_ZERO_EPS) x0 = 1.0f;
    LorentzPoint r(d);
    r.t = x0;
    r.x = x.x;
    return r;
}

bool is_on_manifold(const LorentzPoint& x, float eps) {
    if (x.t <= 0.0f) return false;
    float in = inner(x, x);
    float scale = x.t * x.t;
    if (scale < 1.0f) scale = 1.0f;
    return std::abs(in + 1.0f) < eps * scale;
}

//  batch operations 

void batch_exp_o(const float* Z, int N, int dim, float* X_out) {
    // X_out: N rows of (dim+1) floats [t, x1..xd]
    for (int n = 0; n < N; ++n) {
        const float* z = Z + n * dim;
        float* x_out = X_out + n * (dim + 1);

        float z_norm = 0.0f;
        for (int i = 0; i < dim; ++i) z_norm += z[i] * z[i];
        z_norm = std::sqrt(z_norm);

        if (z_norm < NEAR_ZERO_EPS) {
            x_out[0] = 1.0f;
            std::memset(x_out + 1, 0, dim * sizeof(float));
            continue;
        }
        float theta = numerical::clamp_norm(z_norm);
        float ch = std::cosh(theta);
        float sh = std::sinh(theta);
        x_out[0] = ch;
        for (int i = 0; i < dim; ++i)
            x_out[1 + i] = sh * z[i] / z_norm;
    }
}

void batch_inner_product(const float* Q, int M, const float* X, int N,
                         int dim, float* scores) {
    // Q: M × (dim+1), X: N × (dim+1), scores: M × N
    // ⟨q, x⟩_L = -q₀x₀ + Σqᵢxᵢ
    // Implementation: negate q's time, then dot
    for (int m = 0; m < M; ++m) {
        const float* q = Q + m * (dim + 1);
        float* score_row = scores + m * N;
        float q0_neg = -q[0];
        for (int n = 0; n < N; ++n) {
            const float* x = X + n * (dim + 1);
            float s = q0_neg * x[0];
            for (int i = 0; i < dim; ++i)
                s += q[1 + i] * x[1 + i];
            score_row[n] = s;
        }
    }
}

//  Einstein midpoint 

LorentzPoint einstein_midpoint(const LorentzPoint* points, int n,
                               const float* weights) {
    // Weighted Einstein midpoint in Klein coordinates
    int d = static_cast<int>(points[0].x.size());
    std::vector<float> num(d, 0.0f);
    float denom = 0.0f;

    for (int i = 0; i < n; ++i) {
        float w = (weights != nullptr) ? weights[i] : 1.0f;
        // Klein coordinates: k = x_space / x_time
        float k_norm2 = 0.0f;
        std::vector<float> k(d);
        float x0_clamp = std::max(points[i].t, NEAR_ZERO_EPS);
        for (int j = 0; j < d; ++j) {
            k[j] = points[i].x[j] / x0_clamp;
            k_norm2 += k[j] * k[j];
        }
        float gamma = 1.0f / std::sqrt(std::max(1.0f - k_norm2, NEAR_ZERO_EPS));
        float phi = w * gamma;
        for (int j = 0; j < d; ++j) num[j] += phi * gamma * k[j];
        denom += phi * gamma;
    }

    if (denom < NEAR_ZERO_EPS) {
        LorentzPoint o(d);
        o.t = 1.0f;
        return o;
    }

    float m_k_norm2 = 0.0f;
    std::vector<float> m_k(d, 0.0f);
    for (int j = 0; j < d; ++j) {
        m_k[j] = num[j] / denom;
        m_k_norm2 += m_k[j] * m_k[j];
    }

    float dn = std::sqrt(std::max(1.0f - m_k_norm2, NEAR_ZERO_EPS));
    LorentzPoint result(d);
    result.t = 1.0f / dn;
    for (int j = 0; j < d; ++j) result.x[j] = m_k[j] / dn;
    return project(result);
}

}  // namespace hgnfs::lorentz
